File: serverbanexception.go
package unrealircd

import (
	"errors"
)

// ServerBanException handles server ban exception operations
type ServerBanException struct {
	querier Querier
}

// Add adds a ban exception
func (sbe *ServerBanException) Add(name, exceptionTypes, reason string, setBy, duration *string) (interface{}, error) {
	params := map[string]interface{}{
		"name":             name,
		"exception_types":  exceptionTypes,
		"reason":           reason,
	}
	if setBy != nil {
		params["set_by"] = *setBy
	}
	if duration != nil {
		params["duration_string"] = *duration
	}

	result, err := sbe.querier.Query("server_ban_exception.add", params, false)
	if err != nil {
		return nil, err
	}

	if res, ok := result.(map[string]interface{}); ok {
		if tkl, ok := res["tkl"]; ok {
			return tkl, nil
		}
	}

	return nil, nil
}

// Delete deletes a ban exception
func (sbe *ServerBanException) Delete(name string) (interface{}, error) {
	result, err := sbe.querier.Query("server_ban_exception.del", map[string]interface{}{
		"name": name,
	}, false)
	if err != nil {
		return nil, err
	}

	if res, ok := result.(map[string]interface{}); ok {
		if tkl, ok := res["tkl"]; ok {
			return tkl, nil
		}
	}

	return nil, nil
}

// GetAll returns a list of all exceptions
func (sbe *ServerBanException) GetAll() (interface{}, error) {
	result, err := sbe.querier.Query("server_ban_exception.list", nil, false)
	if err != nil {
		return nil, err
	}

	if res, ok := result.(map[string]interface{}); ok {
		if list, ok := res["list"]; ok {
			return list, nil
		}
	}

	return nil, errors.New("Invalid JSON Response from UnrealIRCd RPC")
}

// Get gets a specific ban exception
func (sbe *ServerBanException) Get(name string) (interface{}, error) {
	result, err := sbe.querier.Query("server_ban_exception.get", map[string]interface{}{
		"name": name,
	}, false)
	if err != nil {
		return nil, err
	}

	if res, ok := result.(map[string]interface{}); ok {
		if tkl, ok := res["tkl"]; ok {
			return tkl, nil
		}
	}

	return nil, nil // not found
}

File: UnrealIRCdRPC/ServerBanException.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Text.Json;

namespace UnrealIRCdRPC
{
    // ServerBanException handles server ban exception operations
    public class ServerBanException
    {
        private readonly IQuerier _querier;

        public ServerBanException(IQuerier querier)
        {
            _querier = querier;
        }

        // Add adds a server ban exception
        public async Task<JsonElement?> AddAsync(string name, string exceptionTypes, string reason, string? setBy = null, string? duration = null)
        {
            var parameters = new Dictionary<string, object>
            {
                { "name", name },
                { "exception_types", exceptionTypes },
                { "reason", reason }
            };
            if (setBy != null) parameters["set_by"] = setBy;
            if (duration != null) parameters["duration_string"] = duration;
            var result = await _querier.QueryAsync("server_ban_exception.add", parameters, false);
            if (result.HasValue && result.Value.ValueKind == JsonValueKind.Object &&
                result.Value.TryGetProperty("tkl", out var tklElement))
            {
                return tklElement;
            }
            return null;
        }

        // Delete deletes a server ban exception
        public async Task<JsonElement?> DeleteAsync(string name)
        {
            var result = await _querier.QueryAsync("server_ban_exception.del", new { name }, false);
            if (result.HasValue && result.Value.ValueKind == JsonValueKind.Object &&
                result.Value.TryGetProperty("tkl", out var tklElement))
            {
                return tklElement;
            }
            return null;
        }

        // GetAll returns a list of all server ban exceptions
        public async Task<JsonElement?> GetAllAsync()
        {
            var result = await _querier.QueryAsync("server_ban_exception.list", null, false);
            if (result.HasValue && result.Value.ValueKind == JsonValueKind.Object &&
                result.Value.TryGetProperty("list", out var listElement))
            {
                return listElement;
            }
            throw new Exception("Invalid JSON Response from UnrealIRCd RPC");
        }

        // Get gets a specific server ban exception
        public async Task<JsonElement?> GetAsync(string name)
        {
            var result = await _querier.QueryAsync("server_ban_exception.get", new { name }, false);
            if (result.HasValue && result.Value.ValueKind == JsonValueKind.Object &&
                result.Value.TryGetProperty("tkl", out var tklElement))
            {
                return tklElement;
            }
            return null; // not found
        }
    }
}